File: src/lib/JsonRpcClient.js
import {URL} from "url";
import axios from "axios";
import shortid from "shortid";

export const JSONRPC_VERSION = "2.0";

/**
 * Create a JSON-RPC client
 * @param {number} port Port of the server
 * @param {string} [host=localhost] Hostname of the server
 * @param {string} [path=/] Path endpoint
 */
export default function JsonRpcClient(port, host, path) {
  host = host || "localhost";
  path = path || "/";
  
  return new Proxy({
    port, host, path,
    version: JSONRPC_VERSION,

    getUrl() {
      return new URL(path, `http://${this.host}:${this.port}`);
    },

    request(method, arg, ...args) {
      var params = [];
      if (args.length > 0) {
        params = [arg].concat(args);
      } else {
        if (typeof arg === "object") {
          params = arg;
        } else {
          params = [arg];
        }
      }

      const url = this.getUrl().toString();
      const data = {
        id: shortid.generate(),
        jsonrpc: this.version,
        method
      };
      if (params) data.params = params;
      return axios.post(url, data).then((resp) => {
        const {result, error} = resp.data;
        if (error) {
          throw Object.assign(new Error(error.message), {
            code: error.code,
            data: error.data
          });
        }
        return result;
      });
    }   
  }, {
    get(target, prop) {
      if (target[prop]) return target[prop];
      return (arg, ...args) => {
        return target.request(prop, arg, ...args);
      };
    }
  });
}

File: controller/jsonrpc_client.py
import requests
from .jsonrpc import JSONRPC_VERSION, ResponseErrorException

# pylint: disable=too-few-public-methods
class JsonRpcClient:
    def __init__(self, port, host='localhost', path='/'):
        self.id = 0
        self.port = port
        self.host = host
        self.path = path
        self.version = JSONRPC_VERSION

    def _create_request(self, method, params):
        self.id += 1
        return {
            'id': self.id,
            'jsonrpc': self.version,
            'method': method,
            'params': params
        }

    def __getattr__(self, name):
        def method(*args, **kwargs):
            url = 'http://%s:%d%s' % (self.host, self.port, self.path)
            params = kwargs if kwargs else args
            json = self._create_request(name, params)
            resp = requests.post(url, json=json)
            json = resp.json()
            try:
                return json['result']
            except KeyError:
                raise ResponseErrorException(json)
        return method
